chain/index/api.go

```diff
@@ -0,0 +1,167 @@
+package index
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+func (si *SqliteIndexer) sanityCheckBackfillEpoch(ctx context.Context, epoch abi.ChainEpoch) error {
+	// should be less than the max non reverted height in the Index
+	var maxNonRevertedHeight sql.NullInt64
+	err := si.stmts.getMaxNonRevertedHeightStmt.QueryRowContext(ctx).Scan(&maxNonRevertedHeight)
+	if err != nil {
+		return xerrors.Errorf("failed to get max non reverted height: %w", err)
+	}
+	if !maxNonRevertedHeight.Valid {
+		return nil
+	}
+	if epoch >= abi.ChainEpoch(maxNonRevertedHeight.Int64) {
+		return xerrors.Errorf("failed to validate index at epoch %d; can only validate at or before epoch %d for safety", epoch, maxNonRevertedHeight.Int64-1)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) validateNullRound(ctx context.Context, epoch abi.ChainEpoch) (*types.IndexValidation, error) {
+	// make sure we do not have ANY non-reverted tipset at this epoch
+	var isNullRound bool
+	err := si.stmts.hasNullRoundAtHeightStmt.QueryRowContext(ctx, epoch).Scan(&isNullRound)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to check if null round exists at height %d: %w", epoch, err)
+	}
+	if !isNullRound {
+		return nil, xerrors.Errorf("index corruption: height %d should be a null round but is not", epoch)
+	}
+
+	return nil, nil
+}
+
+func (si *SqliteIndexer) getTipsetCountsAtHeight(ctx context.Context, height abi.ChainEpoch) (revertedCount, nonRevertedCount int, err error) {
+	err = si.stmts.countTipsetsAtHeightStmt.QueryRowContext(ctx, height).Scan(&revertedCount, &nonRevertedCount)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// No tipsets found at this height
+			return 0, 0, nil
+		}
+		return 0, 0, xerrors.Errorf("failed to query tipset counts: %w", err)
+	}
+
+	return revertedCount, nonRevertedCount, nil
+}
+
+func (si *SqliteIndexer) ChainValidateIndex(ctx context.Context, epoch abi.ChainEpoch, backfill bool) (*types.IndexValidation, error) {
```

Comment:
if we did this all in a single transaction, would that remove the need for a write lock?

Reply:
IIUC, transaction provides atomicity, not exclusivity. Imagine this scenario:

1) `Backfill` is called for epoch e -> it reads tipset `ts` at that epoch
2) `Revert`executes for `ts` and marks the tipset as reverted in `TX1`
 3) `Backfill` executes for `ts` and marks it as applied in `TX2`
 
 `TX1` is committed first and then `TX2` is comitted. 
 
 Basically, it's hard to reason about state here without this lock.

Reply:
@rvagg 

One work-around here it to only allow backfilling for "safe" epochs i.e. `<=head - 30. Then, the rewrite can only ever happen in an edge case.